lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp

```diff
@@ -1092,9 +1096,223 @@ class DecomposeAtenMvOp : public OpRewritePattern<AtenMvOp> {
 };
 } // namespace
 
+// Decompose aten.pixel_shuffle into: aten.permute and aten.reshape operations.
+//
+// If input is a tensor of shape (*leading_dims, C*r*r, H, W), where
+// leading_dims is of size N, then
+//    X = pixel_shuffle(input, upscale_factor)
+//
+// gets replaced with
+//    A = input.reshape(*leading_dims, C, r, r, H, W)
+//    B = A.permute(0, ..., N, N+3, N+1, N+4, N+2)
+//    X = B.reshape(*leading_dims, C, r*H, r*W)
+//
+// 'r' above is referred to as the 'upscale factor' or just 'factor' below.
+namespace {
+class DecomposeAtenPixelShuffleOp
+    : public OpRewritePattern<AtenPixelShuffleOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(AtenPixelShuffleOp op,
+                                PatternRewriter &rewriter) const override {
+
+
+    Location loc = op.getLoc();
+    Value inValue = op.getSelf();
+    auto inType = inValue.getType().cast<BaseTensorType>();
+    auto maybeSizes = inType.getOptionalSizes();
+    if (!maybeSizes) {
+      return rewriter.notifyMatchFailure(
+          op, "Expected input tensor to have known rank.");
+    }
+    auto inShape = maybeSizes.value();
+    auto inRank = inShape.size();
+
+    // TODO: Handle dynamic shapes.
+    //
+    // Currently the decomposition of inputs with dynamic shapes results in a
+    // failure when lowering to linalg. Specifically, the error:
+    //
+    // "Unhandled case of expand/collapse. "
```

Comment:
There is a WIP patch to help with these unhandled dynamic cases (requires some upstream changes to tensor.expand_shape), but it will still generate somewhat non-optimal IR by essentially doing a full collapse -> full expand.

Reply:
Yeah. In a way it is good that the full collapse path isn't in torch-mlir yet -- if it was this PR would have just worked with dynamic shapes and we would have missed the possible better loop nesting.